// @azure/ai-form-recognizer/dist-esm/src/generated/models/index.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
/** Known values of {@link KeyValueType} that the service accepts. */
export var KnownKeyValueType;
(function (KnownKeyValueType) {
    KnownKeyValueType["String"] = "string";
    KnownKeyValueType["SelectionMark"] = "selectionMark";
})(KnownKeyValueType || (KnownKeyValueType = {}));
/** Known values of {@link FieldValueSelectionMark} that the service accepts. */
export var KnownFieldValueSelectionMark;
(function (KnownFieldValueSelectionMark) {
    KnownFieldValueSelectionMark["Selected"] = "selected";
    KnownFieldValueSelectionMark["Unselected"] = "unselected";
})(KnownFieldValueSelectionMark || (KnownFieldValueSelectionMark = {}));
/** Known values of {@link FormLocale} that the service accepts. */
export var KnownFormLocale;
(function (KnownFormLocale) {
    KnownFormLocale["EnAU"] = "en-AU";
    KnownFormLocale["EnCA"] = "en-CA";
    KnownFormLocale["EnGB"] = "en-GB";
    KnownFormLocale["EnIN"] = "en-IN";
    KnownFormLocale["EnUS"] = "en-US";
})(KnownFormLocale || (KnownFormLocale = {}));
/** Known values of {@link FormLanguage} that the service accepts. */
export var KnownFormLanguage;
(function (KnownFormLanguage) {
    KnownFormLanguage["Af"] = "af";
    KnownFormLanguage["Ast"] = "ast";
    KnownFormLanguage["Bi"] = "bi";
    KnownFormLanguage["Br"] = "br";
    KnownFormLanguage["Ca"] = "ca";
    KnownFormLanguage["Ceb"] = "ceb";
    KnownFormLanguage["Ch"] = "ch";
    KnownFormLanguage["Co"] = "co";
    KnownFormLanguage["Crh"] = "crh";
    KnownFormLanguage["Cs"] = "cs";
    KnownFormLanguage["Csb"] = "csb";
    KnownFormLanguage["Da"] = "da";
    KnownFormLanguage["De"] = "de";
    KnownFormLanguage["En"] = "en";
    KnownFormLanguage["Es"] = "es";
    KnownFormLanguage["Et"] = "et";
    KnownFormLanguage["Eu"] = "eu";
    KnownFormLanguage["Fi"] = "fi";
    KnownFormLanguage["Fil"] = "fil";
    KnownFormLanguage["Fj"] = "fj";
    KnownFormLanguage["Fr"] = "fr";
    KnownFormLanguage["Fur"] = "fur";
    KnownFormLanguage["Fy"] = "fy";
    KnownFormLanguage["Ga"] = "ga";
    KnownFormLanguage["Gd"] = "gd";
    KnownFormLanguage["Gil"] = "gil";
    KnownFormLanguage["Gl"] = "gl";
    KnownFormLanguage["Gv"] = "gv";
    KnownFormLanguage["Hni"] = "hni";
    KnownFormLanguage["Hsb"] = "hsb";
    KnownFormLanguage["Ht"] = "ht";
    KnownFormLanguage["Hu"] = "hu";
    KnownFormLanguage["Ia"] = "ia";
    KnownFormLanguage["Id"] = "id";
    KnownFormLanguage["It"] = "it";
    KnownFormLanguage["Iu"] = "iu";
    KnownFormLanguage["Ja"] = "ja";
    KnownFormLanguage["Jv"] = "jv";
    KnownFormLanguage["Kaa"] = "kaa";
    KnownFormLanguage["Kac"] = "kac";
    KnownFormLanguage["Kea"] = "kea";
    KnownFormLanguage["Kha"] = "kha";
    KnownFormLanguage["Kl"] = "kl";
    KnownFormLanguage["Ko"] = "ko";
    KnownFormLanguage["Ku"] = "ku";
    KnownFormLanguage["Kw"] = "kw";
    KnownFormLanguage["Lb"] = "lb";
    KnownFormLanguage["Ms"] = "ms";
    KnownFormLanguage["Mww"] = "mww";
    KnownFormLanguage["Nap"] = "nap";
    KnownFormLanguage["Nl"] = "nl";
    KnownFormLanguage["No"] = "no";
    KnownFormLanguage["Oc"] = "oc";
    KnownFormLanguage["Pl"] = "pl";
    KnownFormLanguage["Pt"] = "pt";
    KnownFormLanguage["Quc"] = "quc";
    KnownFormLanguage["Rm"] = "rm";
    KnownFormLanguage["Sco"] = "sco";
    KnownFormLanguage["Sl"] = "sl";
    KnownFormLanguage["Sq"] = "sq";
    KnownFormLanguage["Sv"] = "sv";
    KnownFormLanguage["Sw"] = "sw";
    KnownFormLanguage["Tet"] = "tet";
    KnownFormLanguage["Tr"] = "tr";
    KnownFormLanguage["Tt"] = "tt";
    KnownFormLanguage["Uz"] = "uz";
    KnownFormLanguage["Vo"] = "vo";
    KnownFormLanguage["Wae"] = "wae";
    KnownFormLanguage["Yua"] = "yua";
    KnownFormLanguage["Za"] = "za";
    KnownFormLanguage["ZhHans"] = "zh-Hans";
    KnownFormLanguage["ZhHant"] = "zh-Hant";
    KnownFormLanguage["Zu"] = "zu";
})(KnownFormLanguage || (KnownFormLanguage = {}));
