// @azure/identity/dist-esm/src/credentials/managedIdentityCredential/appServiceMsi2017.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
import { __awaiter } from "tslib";
import { credentialLogger } from "../../util/logging";
import { msiGenericGetToken } from "./utils";
const logger = credentialLogger("ManagedIdentityCredential - AppServiceMSI 2017");
function expiresInParser(requestBody) {
    // Parse a date format like "06/20/2019 02:57:58 +00:00" and
    // convert it into a JavaScript-formatted date
    return Date.parse(requestBody.expires_on);
}
function prepareRequestOptions(resource, clientId) {
    const queryParameters = {
        resource,
        "api-version": "2017-09-01"
    };
    if (clientId) {
        queryParameters.clientid = clientId;
    }
    return {
        url: process.env.MSI_ENDPOINT,
        method: "GET",
        queryParameters,
        headers: {
            Accept: "application/json",
            secret: process.env.MSI_SECRET
        }
    };
}
export const appServiceMsi2017 = {
    isAvailable() {
        return __awaiter(this, void 0, void 0, function* () {
            const env = process.env;
            return Boolean(env.MSI_ENDPOINT && env.MSI_SECRET);
        });
    },
    getToken(identityClient, resource, clientId, getTokenOptions = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            logger.info(`Using the endpoint and the secret coming form the environment variables: MSI_ENDPOINT=${process.env.MSI_ENDPOINT} and MSI_SECRET=[REDACTED].`);
            return msiGenericGetToken(identityClient, prepareRequestOptions(resource, clientId), expiresInParser, getTokenOptions);
        });
    }
};
